src/fu-engine.c

```diff
@@ -1638,6 +1604,82 @@ fu_engine_install (FuEngine *self,
 					 error);
 }
 
+/**
+ * fu_engine_install_store:
+ * @self: A #FuEngine
+ * @device_id: A device ID
+ * @store: The #AsStore with the firmware metadata
+ * @blob_cab: The #GBytes of the .cab file
+ * @flags: The #FwupdInstallFlags, e.g. %FWUPD_DEVICE_FLAG_UPDATABLE
+ * @error: A #GError, or %NULL
+ *
+ * Installs a specfic firmware file on a device.
+ *
+ * Returns: %TRUE for success
+ **/
+gboolean
+fu_engine_install_store (FuEngine *self,
+			 const gchar *device_id,
+			 AsStore *store,
+			 GBytes *blob_cab,
+			 FwupdInstallFlags flags,
+			 GError **error)
+{
+	FuDevice *device;
+	g_autoptr(GPtrArray) devices = NULL;
+	g_autoptr(GString) str = g_string_new (NULL);
+	gboolean overall = TRUE;
+
+	/* find the specific device */
+	if (g_strcmp0 (device_id, FWUPD_DEVICE_ID_ANY) != 0) {
+		device = fu_device_list_get_by_id (self->device_list, device_id, error);
+		if (device == NULL)
+			return FALSE;
+		return fu_engine_install (self, device, store, blob_cab, flags,
+					  error);
+	}
+
+	devices = fu_device_list_get_all (self->device_list);
+	for (guint i = 0; i < devices->len; i++) {
+		g_autoptr(GError) error_local = NULL;
+		device = g_ptr_array_index (devices, i);
+		if (fu_engine_store_get_app_by_guids (store, device) != NULL) {
+			if (!fu_engine_install (self, device, store, blob_cab,
```

Comment:
Why don't we just bail when the first one fails?

Reply:
If the archive has the same or newer at least you want to keep going. Example I release a dock firmware upgrade for most components but MST is the same.

Reply:
It kinda sucks we're left with `FWUPD_ERROR_INTERNAL` in this case.

Reply:
I don't think so. I think it's FWUPD_ERROR_VERSION_NEWER or FWUPD_ERROR_VERSION_SAME. That's why I'm special casig these errors below.

Reply:
I think it would be better to do the opposite, to check the version before using fu_engine_install(), and then to just abort on any error. I can have a look at this on Tuesday if you like.

Reply:
OK, i've split it up along into a new function that will test specifically things that should be non-fatal when installing from a store of updates.  If the store only has one update or if the ID was specified those will be marked fatal now too.